### ISSUE TITLE
[cmake] Using Protobuf v27 make python pybind11 test failed

### ISSUE THREAD

**Issue by Mizux** [google/or-tools#4307]
# Tests impacted

here the list of impacted tests...

To test:
```sh
cmake -S. -Bbuild -DBUILD_PYTHON=ON -DBUILD_CXX_SAMPLES=OFF -DBUILD_CXX_EXAMPLES=OFF
cmake --build build -j 8
(cd build && ctest -R "python_")
...
96% tests passed, 13 tests failed out of 332

Total Test time (real) = 349.67 sec

The following tests FAILED:
	11 - python_pdlp_pdlp_test (Subprocess aborted)
	14 - python_routing_model_test (Subprocess aborted)
	19 - python_scheduling_rcpsp_test (Subprocess aborted)
	21 - python_math_opt_solver_test (Subprocess aborted)
	170 - python_math_opt_advanced_linear_programming (Subprocess aborted)
	171 - python_math_opt_basic_example (Subprocess aborted)
	172 - python_math_opt_cutting_stock (Subprocess aborted)
	173 - python_math_opt_integer_programming (Subprocess aborted)
	174 - python_math_opt_linear_programming (Subprocess aborted)
	175 - python_math_opt_linear_regression (Subprocess aborted)
	176 - python_math_opt_time_indexed_scheduling (Subprocess aborted)
	178 - python_pdlp_simple_pdlp_program (Subprocess aborted)
	439 - python_python_rcpsp_sat (Subprocess aborted)
Errors while running CTest
Output from these tests are in: /usr/local/google/home/corentinl/work/main/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
```

note: using Protobuf  58baeb4c3b664f8918d24cef5151083d9da9767c everything pass...
Regression seem to have been introduced by http://cl/602725967

# Investigation

Testing one broken test
```sh
cmake -S. -Bbuild -DBUILD_PYTHON=ON -DBUILD_CXX_SAMPLES=OFF -DBUILD_CXX_EXAMPLES=OFF
cmake --build build -j 8
source build/python/venv/bin/activate
python ortools/routing/python/model_test.py

```

Trace:
```
Running tests under Python 3.12.2: /usr/local/google/home/corentinl/work/main/build/python/venv/bin/python
[ RUN      ] ModelTest.testConstantDimensionTSP
[       OK ] ModelTest.testConstantDimensionTSP
[ RUN      ] ModelTest.testCtor
<ortools.routing.python.model.RoutingModel object at 0x7faea8f9d4b0>
[       OK ] ModelTest.testCtor
[ RUN      ] ModelTest.testDimensionTSP
[       OK ] ModelTest.testDimensionTSP
[ RUN      ] ModelTest.testDimensionWithVehicleCapacitiesTSP
[       OK ] ModelTest.testDimensionWithVehicleCapacitiesTSP
[ RUN      ] ModelTest.testDimensionWithVehicleTransitsTSP
[       OK ] ModelTest.testDimensionWithVehicleTransitsTSP
[ RUN      ] ModelTest.testDimensionWithVehicleTransitsVRP
[       OK ] ModelTest.testDimensionWithVehicleTransitsVRP
[ RUN      ] ModelTest.testDisjunctionPenaltyTSP
[       OK ] ModelTest.testDisjunctionPenaltyTSP
[ RUN      ] ModelTest.testDisjunctionTSP
[       OK ] ModelTest.testDisjunctionTSP
[ RUN      ] ModelTest.testMatrixDimensionTSP
[       OK ] ModelTest.testMatrixDimensionTSP
[ RUN      ] ModelTest.testMatrixDimensionVRP
[       OK ] ModelTest.testMatrixDimensionVRP
[ RUN      ] ModelTest.testRoutingLocalSearchFiltering
First solution statistics:
                  | Time (s)
PATH_CHEAPEST_ARC | 3.3e-05
Local search filter statistics (PATH_CHEAPEST_ARC):
                      |     Calls |   Rejects | Time (s) | Rejects/s
VehicleVariableFilter |        12 |         0 | 8.3e-06  |       0
 VariableDomainFilter |        12 |         0 | 7.8e-06  |       0
                Total |        24 |         0 | 1.6e-05  |       0

[       OK ] ModelTest.testRoutingLocalSearchFiltering
[ RUN      ] ModelTest.testRoutingModelParameters
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
F0000 00:00:1720594007.067703 3589844 generated_message_reflection.cc:3620] Check failed: file != nullptr 
*** Check failure stack trace: ***
Fatal Python error: Aborted

Current thread 0x00007faeaa0d0040 (most recent call first):
  File "/usr/local/google/home/corentinl/work/main/ortools/routing/python/model_test.py", line 650 in testRoutingModelParameters
  File "/usr/lib/python3.12/unittest/case.py", line 589 in _callTestMethod
  File "/usr/lib/python3.12/unittest/case.py", line 634 in run
  File "/usr/lib/python3.12/unittest/case.py", line 690 in __call__
  File "/usr/lib/python3.12/unittest/suite.py", line 122 in run
  File "/usr/lib/python3.12/unittest/suite.py", line 84 in __call__
  File "/usr/lib/python3.12/unittest/suite.py", line 122 in run
  File "/usr/lib/python3.12/unittest/suite.py", line 84 in __call__
  File "/usr/lib/python3.12/unittest/runner.py", line 240 in run
  File "/usr/local/google/home/corentinl/work/main/build/python/venv/lib/python3.12/site-packages/absl/testing/_pretty_print_reporter.py", line 82 in run
  File "/usr/lib/python3.12/unittest/main.py", line 281 in runTests
  File "/usr/lib/python3.12/unittest/main.py", line 105 in __init__
  File "/usr/local/google/home/corentinl/work/main/build/python/venv/lib/python3.12/site-packages/absl/testing/absltest.py", line 2653 in _run_and_get_tests_result
  File "/usr/local/google/home/corentinl/work/main/build/python/venv/lib/python3.12/site-packages/absl/testing/absltest.py", line 2689 in run_tests
  File "/usr/local/google/home/corentinl/work/main/build/python/venv/lib/python3.12/site-packages/absl/testing/absltest.py", line 2234 in main_function
  File "/usr/local/google/home/corentinl/work/main/build/python/venv/lib/python3.12/site-packages/absl/app.py", line 254 in _run_main
  File "/usr/local/google/home/corentinl/work/main/build/python/venv/lib/python3.12/site-packages/absl/app.py", line 308 in run
  File "/usr/local/google/home/corentinl/work/main/build/python/venv/lib/python3.12/site-packages/absl/testing/absltest.py", line 2236 in _run_in_app
  File "/usr/local/google/home/corentinl/work/main/build/python/venv/lib/python3.12/site-packages/absl/testing/absltest.py", line 2131 in main
  File "/usr/local/google/home/corentinl/work/main/ortools/routing/python/model_test.py", line 675 in <module>

Extension modules: google._upb._message (total: 1)
zsh: IOT instruction  python ortools/routing/python/model_test.py
```

https://github.com/google/or-tools/blob/f9adb26d1df687ce2362e864ee713f2db00897bc/ortools/routing/python/model_test.py#L648-L651

| Protobuf Version  | Python Test |
| ------------- | ------------- |
| v26.1  | Pass  |
| v27.0  | Failed  |
| v27.1  | Failed  |
| v27.2  | Failed  |

continue to investigate...

ref: https://github.com/protocolbuffers/protobuf/blob/63def39e881afa496502d9c410f4ea948e59490d/src/google/protobuf/generated_message_reflection.cc#L3616-L3620

**Comment by Mizux** [comment]
TLDR: using Protobuf (cmake based build) as a "Static" library will lead to an ODR violation, until now this ODR was without impact/side effect but the Protobuf change this behaviour amd make it crash the runtime.

**Comment by magneticflux-** [comment]
Also seems to break `fzn-cp-sat`:
```
F0801 20:51:14.314146 2793569 generated_message_reflection.cc:3620] Check failed: file != nullptr
```

**Comment by lperron** [comment]
Fixed

**Comment by StefanBruens** [comment]
@lperron - how was this fixed? I hope you don't mean this one:

- https://github.com/google/or-tools/commit/ab029a6f59d775b4fc94e39e1acd611840d34783

**Comment by lperron** [comment]
we dropped it. We build on 29.2 and we forced all libraries to be shared to
remove the dual link static + shared
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le jeu. 2 janv. 2025 à 17:12, StefanBruens ***@***.***> a
écrit :

**Comment by lperron** [comment]
And it took us more than 2 months to make it work.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by StefanBruens** [comment]
So, the actual fix is not to link libraries multiple times, because singletons are no longer singletons then, correct?

Is there actually something in protobuf 29.2 which makes it a hard dependency, or is it sufficient to have a clean setup using single-sourced shared libraries for everything?

Any way, this line is obviously incorrect now:
https://github.com/google/or-tools/blob/82914bc4f9490722acdce4e82adac0f8e11ff967/ortools/python/setup.py.in#L49

**Comment by lperron** [comment]
https://github.com/google/or-tools/blob/bba6615f7de453509129adef9fa1fd564b1068b6/ortools/python/setup.py.in#L49
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le jeu. 2 janv. 2025 à 18:08, StefanBruens ***@***.***> a
écrit :

**Comment by Mizux** [comment]
On v99bugfix (the incoming release v9.12)
https://github.com/google/or-tools/blob/84a35fdc4c4920227083bc928a305137819adfbb/ortools/python/setup.py.in#L49
